### PR TITLE
Fixes #35088 - Remove GCE Compute resource code

### PR DESCRIPTION
**Pull request by stejskalleos**
All code related to GCE compute resource is now in the [Foreman Google plugin](https://github.com/theforeman/foreman_google)

Follow up PRs: 
* https://github.com/theforeman/foreman_google/pull/54
* https://github.com/theforeman/hammer-cli-foreman/pull/606

### CODE REVIEW

File: bundler.d/gce.rb
Comment (stejskalleos) [inline bundler.d/gce.rb:2]: This cause issue while require `google/apis/compute_v1` in `app/models/foreman_google/google_compute.rb`

Comment (stejskalleos) [reply bundler.d/gce.rb:2]: ALso remove app/views/images/form/_gce.html.erb

Comment (adamruzicka) [reply bundler.d/gce.rb:2]: > This cause issue while require

Do we have a workaround for that?

Comment (stejskalleos) [reply bundler.d/gce.rb:2]: Fix: https://github.com/stejskalleos/foreman_google/pull/48

**Comment by adamruzicka** [comment]
@stejskalleos what should we do with this?

**Comment by stejskalleos** [comment]
@adamruzicka this was planned for `3.5`, but was postponed to the `3.6`

**Comment by stejskalleos** [comment]
@adamruzicka Plans changed and we are going to do it in 3.5. Rebased & updated

**Comment by stejskalleos** [comment]
[test katello]

**Comment by stejskalleos** [comment]
@ofedoren Adam is on PTO, could you please do the review and check if it's ready for merge?
@ehelms could check if everything is ok with packaging?

**Review by ofedoren — DISMISSED** [review]
Thanks, @stejskalleos, I'm going to merge this, but I found two places to deal with within stabilization week, so please do:
 - https://github.com/theforeman/foreman/blob/develop/config/application.rb#L59
 - https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/parameters/compute_resource.rb#L23-L28
 
 UPD: But of course I can't just merge until @theforeman/packaging approves.

**Comment by stejskalleos** [comment]
Updated together with https://github.com/theforeman/foreman_google/pull/54

**Comment by Ron-Lavi** [comment]
@theforeman/packaging can you take a look please?

**Comment by ekohl** [comment]
So what is the expected migration path for a user? I opened https://github.com/theforeman/foreman-packaging/pull/8666 to ensure the plugin replaces the core package, but a similar change is needed for the Debian side.

**Comment by stejskalleos** [comment]
[test katello]

**Comment by stejskalleos** [comment]
I'll take care of the deb PR

> So what is the expected migration path for a user?

When upgrading to new version, users who use GCE would have to run `foreman-installer` with `--enable-foreman-google`option

File: app/controllers/api/v2/compute_resources_controller.rb
Comment (adamruzicka) [inline app/controllers/api/v2/compute_resources_controller.rb]: So even though we're moving everything into a plugin, foreman still has to be aware that the plugin exists and do some extra work for it? That feels wrong

File: test/models/compute_resource_test.rb
Comment (adamruzicka) [inline test/models/compute_resource_test.rb:129]: Is the count unstable?

**Review by adamruzicka — COMMENTED** [review]
At first glance, changes proposed here look mostly sensible, will do another round with testing tomorrow.

File: test/models/compute_resource_test.rb
Comment (stejskalleos) [reply test/models/compute_resource_test.rb:129]: Yop, with `twice` it fails on
```
Failure:
ComputeResourceTest#test_0016_.providers returns merge of loaded builtin and registered providers [/home/lstejska/projects/foreman2/app/models/compute_resources/foreman/model/vmware.rb:27]
Minitest::Assertion: unexpected invocation: Fog::Compute.providers()
unsatisfied expectations:
- expected exactly twice, invoked 3 times: Fog::Compute.providers(any_parameters)
satisfied expectations:
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddresses(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddress(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getnames(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getname(any_parameters)
- expected exactly once, invoked once: ComputeResource(id: integer, name: string, description: text, url: string, user: string, password: text, uuid: string, type: string, created_at: datetime, updated_at: datetime, attrs: text, caching_enabled: boolean, domain: string, http_proxy_id: integer).registered_providers(any_parameters)
```

Comment (adamruzicka) [reply test/models/compute_resource_test.rb:129]: And with `.times(3)` instead of `.twice`?

Comment (stejskalleos) [reply test/models/compute_resource_test.rb:129]: Fails with:
```
Failure:
ComputeResourceTest#test_0016_.providers returns merge of loaded builtin and registered providers [/home/lstejska/projects/foreman2/app/models/compute_resources/foreman/model/openstack.rb:23]
Minitest::Assertion: unexpected invocation: Fog::Compute.providers()
unsatisfied expectations:
- expected exactly 3 times, invoked 4 times: Fog::Compute.providers(any_parameters)
satisfied expectations:
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddresses(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getaddress(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getnames(any_parameters)
- allowed any number of times, invoked never: #<AnyInstance:Resolv::DNS>.getname(any_parameters)
- expected exactly once, invoked once: ComputeResource(id: integer, name: string, description: text, url: string, user: string, password: text, uuid: string, type: string, created_at: datetime, updated_at: datetime, attrs: text, caching_enabled: boolean, domain: string, http_proxy_id: integer).registered_providers(any_parameters)


45 tests, 81 assertions, 1 failures, 0 errors, 0 skips

```

**Comment by adamruzicka** [comment]
> I'll take care of the deb PR
> 
> > So what is the expected migration path for a user?
> 
> When upgrading to new version, users who use GCE would have to run `foreman-installer` with `--enable-foreman-google`option

I know I'm sidetracking a bit, but could we do some installer/puppet module magic so it gets enabled automatically on upgrade, but has to be enabled explicitly on new installations to ease the upgrade?

**Comment by ekohl** [comment]
> I know I'm sidetracking a bit, but could we do some installer/puppet module magic so it gets enabled automatically on upgrade, but has to be enabled explicitly on new installations to ease the upgrade?

IMHO that's very relevant. First of all, packaging side I'm trying to make sure the foreman_google package replaces the `foreman-gce` package. That means a `dnf/apt upgrade` will mean users who have it installed today, will continue to have it installed. Even if they don't use the installer:

Then on the installer side I'm asking for a similar thing. Additionally, when both the CLI and the Foreman Google plugin are enabled, it should also enabled the CLI Google plugin.

I think this PR should be merged together with the following PRs:

* https://github.com/theforeman/foreman-packaging/pull/8666 (RPM packaging)
* https://github.com/theforeman/foreman-packaging/pull/8667 (deb packaging)
* https://github.com/theforeman/foreman-installer/pull/758 installer

**Comment by stejskalleos** [comment]
@ekohl installer and deb updated, however I don't if I should be one doing the deb packaging, I don't understand it and if we want to have it in 3.5 maybe someone from packaging should take care of it.

**Comment by adamruzicka** [comment]
Foreman part works well from what I've seen, however there are some issues with hammer.

**Comment by stejskalleos** [comment]
> Foreman part works well from what I've seen, however there are some issues with hammer.

See my https://github.com/stejskalleos/hammer-cli-foreman-google/pull/1#issuecomment-1302132811 in hammer plugin, for me it works well, maybe you didn't have correct branches everywhere?

**Review by ekohl — APPROVED** [review]
While normally this is not really something I'd do during stabilization, I'm going to pull the trigger and merge this if @adamruzicka also approves it. Then I'll also merge packaging and installer.

**Review by adamruzicka — APPROVED** [review]
> See my https://github.com/stejskalleos/hammer-cli-foreman-google/pull/1#issuecomment-1302132811 in hammer plugin, for me it works well, maybe you didn't have correct branches everywhere?

It was something on my end. Now that it is resolved, I'd say this is good to go

**Pull request merged**
